### PR TITLE
Multi-window command routing

### PR DESCRIPTION
**Pull request by rolfheij-sil**
> [!IMPORTANT]
> **Superseded by #2621.** This PR is kept open, as a draft, as the record of the original hackathon
> spike. The work continues in [#2621](https://github.com/paranext/paranext-core/pull/2621), which
> starts from these commits, rebases them onto current main, and carries them forward — review there,
> not here.
>
> What changed in the successor: the branch was rebased across ~330 commits of main; the remaining
> open items listed below were addressed; the feature was actually run with two and three windows,
> which found a multi-window quit skipping the shutdown Send/Receive and a set of event-emitter and
> web-view-id collisions that made the second window inert. Ticket:
> [PT-1891](https://paratextstudio.atlassian.net/browse/PT-1891).

---

## Summary
- Adds window-scoped service registrations so multiple renderer windows can coexist without name collisions
- Introduces focus-based routing proxies in the main process that forward commands and WebViewService calls to the focused window
- Adds keyboard shortcut (Ctrl+Shift+N / Cmd+Shift+N) and menu item for opening new windows
- Fixes extension host `papi.window` access, `platform.about` command scoping, `openWebView` cross-window search, and localStorage migration safety

## How it works
Each renderer registers under scoped names (e.g. `WebViewService-1`, `platform.openSettings-1`). The main process registers proxies under the generic names that query `focusedWindowId` and forward to the correct window's scoped handler.

```
Extension Host → sendCommand('platform.openSettings')
    → Main Process proxy (generic name)
    → queries focusedWindowId (= 2)
    → forwards to 'platform.openSettings-2'
    → Window 2's renderer handles it
```

### New files
| File | Purpose |
|------|---------|
| `src/main/services/window-state.service.ts` | Tracks focused window ID and windows array |
| `src/main/services/web-view-routing.service.ts` | WebViewService proxy that routes to focused window, with cross-window search for existing webviews |
| `src/main/services/command-routing.service.ts` | Command proxies for 8 renderer-hosted commands (including `platform.about`) |

### Modified files
| File | Change |
|------|--------|
| `src/main/main.ts` | Focus tracking, per-window keyboard shortcuts (Ctrl+Shift+N), startup flag scoping, startup wiring |
| `src/renderer/services/web-view.service-host.ts` | Scoped network object + command registration |
| `src/renderer/services/window.service-host.ts` | Scoped data provider registration |
| `src/renderer/services/dialog.service-host.ts` | Scoped `platform.about` command registration |
| `src/renderer/services/localStorage.service.ts` | Per-window localStorage with legacy migration (preserves legacy keys) |
| `src/shared/services/web-view.service-model.ts` | `RENDERER_HOSTED_COMMAND_NAMES` constant (includes `platform.about`) |
| `src/shared/services/window.service.ts` | Extension host support via focused window resolution |
| `src/shared/services/window.service-model.ts` | Updated JSDoc for multi-window |
| `src/declarations/papi-shared-types.ts` | `platform.createWindow` and `platform.getFocusedWindowId` command types |

### Removed files
| File | Reason |
|------|--------|
| `src/main/services/window-aggregator.service-host.ts` | Unused scaffolding |
| `src/shared/services/window-aggregator.service-model.ts` | Unused scaffolding |
| `src/shared/services/window-aggregator.service.ts` | Unused scaffolding |
| `src/shared/models/window-controller.interface.ts` | Unused scaffolding |

## Remaining open items

These were identified during review and are not yet addressed:

- **Incomplete localStorage migration** — 4 renderer files still use raw `localStorage` instead of `localWindowStorage`
- **Menu item group** — "Open new window" should move from `platform.projectSettings` to `platform.windowGroup`
- **Test coverage** — No tests for localStorage migration, command routing, or web view routing
- **`windowCloseUnsubscribers` fragility** — Should wrap each unsubscriber in try/catch
- **Double type assertion** — `as unknown as IWindowService` in `window.service-host.ts`
- **Startup validation** — `RENDERER_HOSTED_COMMAND_NAMES` should validate handlers are registered

## Test plan
- [ ] `npm run build` succeeds
- [ ] `npm test` passes
- [ ] `npm run lint` passes
- [ ] `npm run typecheck` passes
- [ ] Window 1 loads normally with `npm start`
- [ ] Open second window via Platform menu → "Open new window"
- [ ] Open second window via Ctrl+Shift+N / Cmd+Shift+N
- [ ] Open a project in Window 2's Home → opens **in Window 2** (not Window 1)
- [ ] Open a project in Window 1 → opens in Window 1
- [ ] Settings command in Window 2 → opens in Window 2
- [ ] About dialog works in both windows
- [ ] Close Window 2 → Window 1 remains functional
- [ ] Close Window 1 (with Window 2 open) → Window 2 takes over
- [ ] Second window opens in normal state (not minimized/maximized) even if app was started with `--maximize`

🤖 Generated with [Claude Code](https://claude.ai/code)

---

# Code Review Summary

**Branch**: multi-window

**Base**: origin/main

**Date**: 2026-04-10

**Review model**: Claude Opus 4.6

**Files changed**: 22

## Overview

This PR adds multi-window support to Platform.Bible, allowing users to open multiple Electron BrowserWindows. The core approach introduces per-window scoping for renderer-hosted commands and services: each renderer registers under a scoped name (e.g., `WebViewService-1`, `platform.openSettings-1`), and the main process registers proxy commands/services under generic names that route to the focused window. New infrastructure includes a window state tracker (`window-state.service.ts`), command routing proxies (`command-routing.service.ts`), web view routing across windows (`web-view-routing.service.ts`), and a `localWindowStorage` wrapper that namespaces localStorage keys per window with migration for legacy unprefixed keys. Two new PAPI commands are exposed: `platform.createWindow` and `platform.getFocusedWindowId`. A keyboard shortcut (Ctrl+Shift+N / Cmd+Shift+N) and menu item ("Open new window") are provided.

## API Changes

- `papi-shared-types` (`CommandHandlers`): Added `'platform.createWindow': () => Promise<void>` command
- `papi-shared-types` (`CommandHandlers`): Added `'platform.getFocusedWindowId': () => Promise<number | undefined>` command
- `shared/global-this.model` (declared in `papi.d.ts`): Added `var windowId: string | null` global variable
- `shared/data/platform.data` (declared in `papi.d.ts`): Added `WINDOW_ID` constant export
- `shared/services/web-view.service-model` (declared in `papi.d.ts`): Added `RENDERER_HOSTED_COMMAND_NAMES` constant export (readonly tuple, includes `platform.about`)
- `shared/models/project-lookup.service-model` (declared in `papi.d.ts`): Fixed typo in JSDoc — "for then network" to "for the network"
- New module: `localWindowStorage` default export from `src/renderer/services/localStorage.service.ts`
- New menu item: "Open new window" in application menu with Ctrl+Shift+N / Cmd+Shift+N shortcut

### Critical — Must address before merge

- [x] **`papi.window` service now always fails in the extension host process** _(fixed during review: when windowId is absent, resolve focused window's scoped provider via `platform.getFocusedWindowId` command instead of rejecting. Also fixed promise-constructor anti-pattern.)_

- [x] **`platform.about` command not scoped for multi-window** _(fixed during review: added `platform.about` to `RENDERER_HOSTED_COMMAND_NAMES` and scoped registration in `dialog.service-host.ts`)_

- [x] **`openWebView` proxy doesn't search across windows for `existingId`** _(fixed during review: when `existingId` is provided, search all windows via `findOwnerService` helper before falling back to focused window)_

- [x] **Legacy localStorage migration may permanently lose data when Electron window ID changes across restarts** _(fixed during review: removed `localStorage.removeItem(key)` so legacy key stays as permanent fallback)_

Author response: Author confirmed `platform.about` and `openWebView` issues were overlooked. Author confirmed extension host case was not considered. Author does not know whether Electron's `BrowserWindow.id` is stable across restarts — fix preserves the legacy key as a safety measure regardless.

### Important — Should address before merge

- [ ] **Incomplete localStorage migration** — `theme.service-host.ts`, `scroll-group.service-host.ts`, `platform-bible-toolbar.tsx`, and `use-recent-scripture-refs.hook.ts` still call `localStorage` directly instead of `localWindowStorage`. In multi-window, these would share state across windows unintentionally.

- [ ] **"Open new window" menu item placed in wrong group** — Currently in `platform.projectSettings` group alongside "Settings". A dedicated `platform.windowGroup` already exists and would be more semantically appropriate. The `localizeNotes` also reference "Platform" which doesn't match the actual group.

- [ ] **No tests for localStorage migration logic** — The `localWindowStorage` service contains migration logic that should have unit tests covering: prefixed values returned when present, legacy unprefixed values migrated correctly, null returned when neither key exists.

- [ ] **No tests for multi-window routing logic** — `reloadWebView`, `getOpenWebViewDefinition`, `openWebView`, and the command routing proxy contain non-trivial fan-out logic with subtle edge cases that should have unit test coverage.

- [ ] **`windowCloseUnsubscribers` cleanup could be fragile** — If one unsubscriber throws, the remaining ones won't be called. Should wrap each call in try/catch.

- [ ] **Double type assertion in `window.service-host.ts`** — `as unknown as IWindowService` is a code smell; consider improving the type-safe registration path.

- [ ] **`RENDERER_HOSTED_COMMAND_NAMES` routing lacks startup validation** — If a name is added to the array but the renderer doesn't register a handler, the proxy silently fails. Should add a startup assertion.

- [x] ~~**Localization key out of alphabetical order**~~ _(fixed during review: moved `%mainMenu_openNewWindow%` to correct position)_

- [x] ~~**DRY violation in "find owner window" logic**~~ _(fixed during review: extracted `findOwnerService` helper used by `reloadWebView`, `getOpenWebViewDefinition`, and `openWebView`)_

- [x] ~~**`window.service.ts` uses promise-constructor anti-pattern**~~ _(fixed during review: replaced with direct async IIFE)_

Author response: Author confirmed all open items should be addressed. localStorage migration for remaining files should be done. Menu item should move to `platform.windowGroup`.

### Minor — Consider

- [x] **`createWindow()` not awaited in command handler** _(fixed during review: added `await`)_
- [x] **`createWindow()` not error-handled in keyboard shortcut** _(fixed during review: added `.catch()` with logging)_
- [x] **`START_MINIMIZED`/`--maximize` flags affect all new windows** _(fixed during review: wrapped with `isFirstWindow` check)_
- [x] **`getFocusedWindowId` result schema mismatch** _(fixed during review: changed to `oneOf: [number, null]`)_
- [x] **"Open New Window" uses title case** _(fixed during review: changed to "Open new window")_
- [x] **Storybook terminology guide says "only one window"** _(fixed during review: updated to reflect multi-window support)_
- [x] **Stale "main window" JSDoc throughout** _(fixed during review: updated zoomIn/zoomOut JSDoc, PAPI command summaries, and IWindowService method docs)_
- [x] **`WINDOW_ID` JSDoc is vague** _(fixed during review: changed to "Query parameter key used to pass the Electron BrowserWindow ID to the renderer process")_
- [x] **`const windows = getWindows()` misleading at module scope** _(fixed during review: added comment clarifying it's a live reference)_
- [x] **No keyboard shortcut for "Open new window"** _(fixed during review: added Ctrl+Shift+N / Cmd+Shift+N)_
- [x] **Unused window aggregator and window controller scaffolding** _(fixed during review: removed 4 files and all references)_
- [ ] ~~**`focusedWindowSubscribers` Set grows without cleanup**~~ _(dismissed: analysis agent hallucinated; the actual `windowServiceCache` and `windowServicePending` Maps have proper cleanup via `onDidDispose` and window close handlers)_
- [ ] ~~**No JSDoc on `command-routing.service.ts` exported functions**~~ _(dismissed: both files already have JSDoc on all exported functions)_
- [ ] ~~**Menu item correctly omits ellipsis**~~ _(positive observation, not an issue)_

Author response: Author confirmed no remaining minor items need discussion.

#### Shared Regions Modified

None.

#### Extension Config Changes

None.

## Positive Observations

- The routing proxy pattern (main process registers generic names, renderers register scoped names) is a clean approach that avoids breaking the existing single-window API contract
- The `RENDERER_HOSTED_COMMAND_NAMES` constant with runtime validation in `web-view.service-host.ts` is an excellent defensive pattern that catches configuration drift during development
- The `windowServiceCache` with `onDidDispose` cleanup and in-flight deduplication (`windowServicePending`) avoids stale references while preventing redundant network lookups on high-frequency input events
- Good use of `Promise.allSettled` in `getOpenWebViewDefinition` to query all windows without one failure blocking the others
- The zoom factor subscription was correctly refactored from global to per-window with proper cleanup
- macOS menubar subscription was correctly moved to global scope with proper cleanup on `will-quit`
- The `papi.d.ts` was regenerated in a dedicated commit rather than hand-edited
- Good separation of concerns: window state, command routing, and web view routing are in separate service files
- The typo fix in `project-lookup.service-model.ts` is a welcome cleanup

## Interview Notes

The author confirmed the purpose is to add multi-window support to Platform.Bible. Several critical issues from Devin's automated review were discussed and subsequently fixed:

- The `platform.about` scoping issue and `openWebView` cross-window search were both confirmed as oversights — both fixed during review
- The extension host regression for `papi.window` was not considered during implementation — fixed by resolving the focused window's scoped provider via command
- The localStorage migration data loss risk (unstable window IDs) was addressed by not deleting the legacy key

Additional issues found by the code review agents were also fixed: `START_MINIMIZED`/`--maximize` flags now only apply to the first window, stale "main window" JSDoc was updated throughout, and `createWindow()` error handling was added.

The author made clear decisions during the review: unused scaffolding (window aggregator service, window controller interface) should be removed rather than kept as placeholders.

## In-Review Quality Check

All checks passed after in-review changes:

| Check | Result |
|-------|--------|
| `build:types` | Passed — regenerated `papi.d.ts` |
| `typecheck` | Passed — all 15 projects clean |
| `lint` | Passed — no errors |
| `format` | Passed |
| `test` | Passed — 944 tests across all workspaces (platform-bible-react skipped due to pre-existing missing Playwright binaries) |

## Suggested Review Focus

Prioritized areas for the author-reviewer meeting:

- [ ] **Incomplete localStorage migration** — 4 renderer files still use raw `localStorage` and will share state across windows
- [ ] **Menu item group** — "Open new window" should move from `platform.projectSettings` to `platform.windowGroup`
- [ ] **Test coverage** — No tests for localStorage migration, command routing, or web view routing
- [ ] **`windowCloseUnsubscribers` fragility** — Should wrap each unsubscriber in try/catch
- [ ] **Extension host window service caching** — The focused window's provider is resolved once and cached; subsequent calls from the extension host go to the same window even if focus changes. Acceptable for now but worth discussing.
- [ ] **Double type assertion** — `as unknown as IWindowService` in `window.service-host.ts`
- [ ] **Startup validation** — `RENDERER_HOSTED_COMMAND_NAMES` routing should validate that handlers are actually registered

- - -
This change is [Reviewable](https://reviewable.io/reviews/paranext/paranext-core/2051)

[PT-1891]: https://paratextstudio.atlassian.net/browse/PT-1891?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

File: src/renderer/services/web-view.service-host.ts [hunk 1]
```diff
@@ -1821,9 +1828,27 @@ export async function startWebViewService(): Promise<void> {
...
+  const extra = routedCommands.filter((k) => !handlerKeys.includes(k));
+  if (missing.length > 0) {
+    const msg = `Commands registered in renderer but not in RENDERER_HOSTED_COMMAND_NAMES: ${missing.join(', ')}`;
+    if (!globalThis.isPackaged) throw new Error(msg);
+    logger.warn(msg);
+  }
+  if (extra.length > 0) {
+    const msg = `Commands in RENDERER_HOSTED_COMMAND_NAMES but not registered in renderer: ${extra.join(', ')}`;
+    if (!globalThis.isPackaged) throw new Error(msg);
+    logger.warn(msg);
```
[the remark this answers is not in the document]
Comment (rolfheij-sil) [reply src/renderer/services/web-view.service-host.ts]: This is a false positive. The validation has two separate paths:

1. **`missing`** (L1836-1839): commands in `commandHandlers` but NOT in `RENDERER_HOSTED_COMMAND_NAMES` → throws in dev. This catches untracked local handlers.
2. **`extra`** (L1843-1846): commands in `RENDERER_HOSTED_COMMAND_NAMES` but NOT in `commandHandlers` → `logger.debug()` only. No throw.

`platform.about` falls into the `extra` bucket, which only logs. The comment at L1841-1842 explicitly documents this:

> *RENDERER_HOSTED_COMMAND_NAMES may include commands registered in other renderer files (e.g. platform.about in dialog.service-host.ts), so we only warn about extras rather than throw*

No crash, no code change needed.

File: src/renderer/services/localStorage.service.ts [hunk 2]
```diff
@@ -0,0 +1,35 @@
...
+    // Migration: check for legacy unprefixed key from before multi-window support.
+    // Copy to the prefixed key but do NOT delete the legacy key — Electron's BrowserWindow.id
+    // is not guaranteed to be stable across restarts, so the legacy key must remain as a fallback.
+    const legacyValue = localStorage.getItem(key);
+    // localStorage.getItem returns null when the key doesn't exist
+    // eslint-disable-next-line no-null/no-null
+    if (legacyValue !== null) {
+      localStorage.setItem(prefixedKey, legacyValue);
+      return legacyValue;
+    }
```
[the remark this answers is not in the document]
Comment (rolfheij-sil) [reply src/renderer/services/localStorage.service.ts:25]: Thanks for flagging. We intentionally keep the legacy unprefixed key as a fallback (per the earlier review comment discussion that led to commit d2e1ecaec6) because `BrowserWindow.id` is not stable across Electron restarts — deleting it, or gating it behind a "migration done" marker, would reintroduce the permanent-data-loss bug that earlier review correctly caught.

A newly-opened second window was never intended to inherit window 1's *current* layout — the windows are independent, so falling back to the pre-multi-window legacy layout as that window's default is the desired behavior, not a regression. The repeated `setItem` on fallback is idempotent and harmless.

Closing as working-as-intended.

**Comment by rolfheij-sil** [comment]
Superseded by https://github.com/paranext/paranext-core/pull/2621
